### DISCUSSION TITLE
MAL CounterWindow may should use `metric name` as  ID name instead of `Sample name`.

### DISCUSSION THREAD

**Discussion by weixiang1862** [apache/skywalking#11522] [General]
I have three mal rules, they all use sample `nginx_http_requests_total` to analyse metrics:
```
- name: http_requests
  exp: nginx_http_requests_total.sum(['service']).rate('PT1M')
- name: http_4xx_requests
  exp: nginx_http_requests_total.tagMatch("status", "400|401|403|404|405").sum(['service']).rate('PT1M')
- name: http_5xx_requests
  exp: nginx_http_requests_total.tagMatch("status", "500|502|503|504").sum(['service']).rate('PT1M')
```
```
# HELP nginx_http_requests_total Number of HTTP requests
# TYPE nginx_http_requests_total counter
nginx_http_requests_total{status="200",route="/test/**"} 197
nginx_http_requests_total{status="404",route="/test_404/**"} 26
nginx_http_requests_total{status="500",route="/test_500/**"} 47
```

MAL use `CounterWindow` which record history smaple value to calculate rate, but the above three mal rules share the same metric name as ID key, it will cause their history value effect each other.

In order to isolate metrics history smaple value, I think metric name is better for `CounterWindow` ID name: 

![image](https://github.com/apache/skywalking/assets/16346644/2f7025c0-1b51-4378-8a58-f0267c1a6c71)

**Comment by wu-sheng** [comment]
The `labels` of the ID object are a part of `EqualsAndHashCode`, and include the label keys and values. 

If those are the same, they are the same original metrics, we don't need to cache them in the windows duplicated.

**Comment by weixiang1862** [comment]
Yes , metric should share countwindow cache if the has same id name and label.

But in my case, `meter_nginx_service_http_requests`, `meter_nginx_service_http_4xx_requests`, `meter_nginx_service_http_4xx_requests`, they are independent metrics and should not share the same cache.

If they share the same cache, It will cause `http_requests (HTTP Request Trend)` metric can not fall back to 0 even if there are not http calls for a long time, because `http_requests` rate caculation effected by `http_4xx_requests` and `http_5xx_requests`：

![image](https://github.com/apache/skywalking/assets/16346644/f5ea00d6-040c-41f1-b4f8-c4b44d9a8c95)

They share the same cache due to the same ID key: 
```
ID(name=nginx_http_requests_total, labels={service=nginx::portal})
```

I think they need three independent ID key:
```
ID(name=meter_nginx_service_http_requests, labels={service=nginx::portal})
ID(name=meter_nginx_service_http_4xx_requests, labels={service=nginx::portal})
ID(name=meter_nginx_service_http_4xx_requests, labels={service=nginx::portal})
```

**Reply by wu-sheng** [reply]
OK, that seems to make sense. Are you going to submit a pull request? We could review more details on the code changes.

**Reply by weixiang1862** [reply]
Yes, I will submit it.

**Comment by wu-sheng** [comment]
Update, codes are on https://github.com/apache/skywalking/pull/11526. Have been merged.
